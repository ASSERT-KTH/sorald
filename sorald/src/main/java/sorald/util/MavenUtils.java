package sorald.util;

import java.nio.file.Path;
import java.util.List;
import spoon.MavenLauncher;

/** Utility methods for working with Maven projects */
public class MavenUtils {
    private MavenUtils() {}

    /**
     * Resolve the full classpath for a Maven project.
     *
     * @param source Path to the root of a Maven project
     * @return The full source classpath
     */
    public static List<String> resolveClasspath(Path source) {
        MavenLauncher launcher =
                new MavenLauncher(source.toString(), MavenLauncher.SOURCE_TYPE.ALL_SOURCE);
        return List.of(launcher.getEnvironment().getSourceClasspath());
    }

    /**
     * Test whether or not the given source path points to the root of a Maven project.
     *
     * @param source A path
     * @return true iff the source path points to the root of a Maven project
     */
    public static boolean isMavenProjectRoot(Path source) {
        return source.resolve("pom.xml").toFile().isFile();
    }
}
