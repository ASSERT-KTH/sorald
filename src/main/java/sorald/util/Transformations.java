package sorald.util;

import spoon.reflect.code.CtExpression;
import spoon.reflect.code.CtUnaryOperator;
import spoon.reflect.code.UnaryOperatorKind;

/** Reusable transformations for use across the entire project. */
public class Transformations {

    private Transformations() {}

    /**
     * Negate an expression, i.e. turn `expr` into `!expr`.
     *
     * @param expr Expression to negate
     * @param <T> Type of the expression
     * @return The negation of the input expression
     */
    public static <T> CtUnaryOperator<T> not(CtExpression<T> expr) {
        CtUnaryOperator<T> operator = expr.getFactory().createUnaryOperator();
        operator.setKind(UnaryOperatorKind.NOT);
        operator.setOperand(expr);
        return operator;
    }
}
