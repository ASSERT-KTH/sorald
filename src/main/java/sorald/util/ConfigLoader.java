package sorald.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Paths;
import java.util.Properties;

public class ConfigLoader {
    private static final String CONFIG_FILE_NAME = "config.properties";
    private static final File CONFIG_FILE =
            Paths.get("src")
                    .resolve("main")
                    .resolve("resources")
                    .resolve(CONFIG_FILE_NAME)
                    .toFile();
    private static final Properties properties = loadProperties();

    private ConfigLoader() {
        throw new IllegalStateException("Utility class cannot be instantiated.");
    }

    private static Properties loadProperties() {
        Properties configuration = new Properties();
        try (InputStream inputStream = new FileInputStream(CONFIG_FILE)) {
            configuration.load(inputStream);
            return configuration;
        } catch (IOException ignore) {
            throw new RuntimeException("Could not read config file"); // NOSONAR:S112
        }
    }

    /** Fetches URL of SonarJavaPlugin from config. */
    public static String getSonarJavaPluginUrl() {
        return properties.getProperty("SONAR_JAVA_PLUGIN_URL");
    }
}
