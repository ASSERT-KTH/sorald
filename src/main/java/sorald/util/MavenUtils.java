package sorald.util;

import java.nio.file.Path;
import java.util.List;
import spoon.MavenLauncher;

/** Utility methods for working with Maven projects */
public class MavenUtils {
    private MavenUtils() {}

    /**
     * Resolve the full classpath for a Maven project.
     *
     * @param source Path to the root of a Maven project
     * @return The full source classpath
     * @throws IllegalArgumentException If source does not point to the root of a Maven project
     */
    public static List<String> resolveClasspath(Path source) {
        if (!isMavenProjectRoot(source)) {
            throw new IllegalArgumentException(
                    String.format("'%s' is not a Maven project root", source));
        }

        MavenLauncher launcher =
                new MavenLauncher(source.toString(), MavenLauncher.SOURCE_TYPE.ALL_SOURCE);
        return List.of(launcher.getEnvironment().getSourceClasspath());
    }

    private static boolean isMavenProjectRoot(Path source) {
        return source.resolve("pom.xml").toFile().isFile();
    }
}
