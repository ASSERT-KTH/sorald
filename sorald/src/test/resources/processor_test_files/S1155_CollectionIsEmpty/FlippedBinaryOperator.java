import java.util.Collection;

public class FlippedBinaryOperator {
    public static void main(String[] args) {
        Collection myCollection = null;

        if (0 == myCollection.size()) {  // Noncompliant
            /* ... */
        }
        if (0 >= myCollection.size()) {  // Noncompliant
            /* ... */
        }
    }
}
