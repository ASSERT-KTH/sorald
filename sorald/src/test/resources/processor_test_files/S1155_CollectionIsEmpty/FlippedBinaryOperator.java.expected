import java.util.Collection;

public class FlippedBinaryOperator {
    public static void main(String[] args) {
        Collection myCollection = null;

        if (myCollection.isEmpty()) {  // Noncompliant
            /* ... */
        }
        if (myCollection.isEmpty()) {  // Noncompliant
            /* ... */
        }
    }
}
